lib/grandpa/finalisation_integration_test.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+//go:build integration
+
+package grandpa
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_finalisationHandler_Stop_ShouldHalt_Services(t *testing.T) {
+	t.Parallel()
+
+	testcases := map[string]struct {
+		wantErr    error
+		errString  string
+		newHandler func(*gomock.Controller) *finalisationHandler
+	}{
+		"halt_ephemeral_services_after_calling_stop": {
+			// when we start the finalisation handler we instantiate
+			// and call the Run method from each ephemeral services
+			// (votingHandler, finalisationEngine) since they are mocked
+			// they will wait until the Stop method being called to release
+			// the blocking channel and return from the function
+			newHandler: func(ctrl *gomock.Controller) *finalisationHandler {
+				builder := func() (engine ephemeralService, voting ephemeralService) {
+					engineStopCh := make(chan struct{})
+					mockEngine := NewMockephemeralService(ctrl)
+
+					mockEngine.EXPECT().Run().DoAndReturn(func() error {
+						<-engineStopCh
+						return nil
+					})
+					mockEngine.EXPECT().Stop().DoAndReturn(func() error {
+						close(engineStopCh)
+						return nil
+					})
+
+					votingStopCh := make(chan struct{})
+					mockVoting := NewMockephemeralService(ctrl)
+					mockVoting.EXPECT().Run().DoAndReturn(func() error {
+						<-votingStopCh
+						return nil
+					})
+					mockVoting.EXPECT().Stop().DoAndReturn(func() error {
+						close(votingStopCh)
+						return nil
+					})
+					return mockEngine, mockVoting
+				}
+
+				return &finalisationHandler{
+					newServices: builder,
+					// mocked initiate round function
+					initiateRound: func() error { return nil },
+					stopCh:        make(chan struct{}),
+					handlerDone:   make(chan struct{}),
+				}
+			},
+		},
+		"halt_fails_to_stop_one_ephemeral_service": {
+			wantErr:   errServicesStopFailed,
+			errString: "services stop failed: cannot stop finalisation engine test",
+			newHandler: func(ctrl *gomock.Controller) *finalisationHandler {
+				builder := func() (engine ephemeralService, voting ephemeralService) {
+					engineStopCh := make(chan struct{})
+					mockEngine := NewMockephemeralService(ctrl)
```

Comment:
Are mocked tests part of integration tests just because they are long running?

Reply:
This is really a unit test, but we need to wait some time (here 2s) for:

- ephemeral services to be set (nil initially)
- goroutines to be launched so mock Run methods are called as expected

Now ephemeral services can be (and should be) set in the `Start` method before we call `run` in a goroutine; and for the goroutines Run method expectations, we can just add `.MaxTimes(1)` to the Run expectations. Then that test is instant and can be moved to a unit test.

Pushed changes to https://github.com/ChainSafe/gossamer/pull/2938/files